components/engine_traits/src/util.rs

```diff
@@ -21,3 +24,151 @@ pub fn check_key_in_range(
+    target: PathBuf,
+    target_staging: PathBuf,
+}
+
+impl RaftDataStateMachine {
+    pub fn new(source: &str, target: &str) -> Self {
+        let source = PathBuf::from(source);
+        let source_staging = source.with_extension("STAGING");
+        let target = PathBuf::from(target);
+        let target_staging = target.with_extension("STAGING");
+        Self {
+            source,
+            source_staging,
+            target,
+            target_staging,
+        }
+    }
+
+    /// Verifies if invariants still hold.
+    pub fn validate(&self, should_exist: bool) -> std::result::Result<(), String> {
+        if Self::data_exists(&self.source) && Self::data_exists(&self.target) {
+            return Err(format!(
+                "Found multiple raft data sets: {}, {}",
+                self.source.display(),
+                self.target.display()
+            ));
+        }
+        if Self::data_exists(&self.source_staging) && Self::data_exists(&self.target_staging) {
+            return Err(format!(
+                "Found multiple raft data sets: {}, {}",
+                self.source_staging.display(),
+                self.target_staging.display()
+            ));
+        }
+        let exists = Self::data_exists(&self.source_staging)
+            || Self::data_exists(&self.target_staging)
+            || Self::data_exists(&self.source)
+            || Self::data_exists(&self.target);
+        if exists != should_exist {
+            if should_exist {
+                return Err("Cannot find raft data set.".to_owned());
+            } else {
+                return Err("Found raft data set when it should not exist.".to_owned());
+            }
+        }
+        Ok(())
+    }
+
+    pub fn before_open_target(&mut self) -> Option<PathBuf> {
+        if Self::data_exists(&self.target_staging) {
+            assert!(!Self::data_exists(&self.source_staging));
+            Self::remove_dir_safe(&self.source);
+            Self::rename_dir_safe(&self.target_staging, &self.target);
+            return None;
+        }
+        if Self::data_exists(&self.source_staging) {
+            Self::remove_dir_safe(&self.source);
+        } else if Self::data_exists(&self.source) {
+            Self::rename_dir_safe(&self.source, &self.source_staging);
+        } else {
+            return None;
+        }
+        Some(self.source_staging.clone())
+    }
+
+    pub fn after_dump_data(&mut self) {
+        assert!(Self::data_exists(&self.target));
+        assert!(!Self::data_exists(&self.source));
+        Self::remove_dir_safe(&self.source_staging);
+        Self::remove_dir_safe(&self.target_staging);
+    }
+
+    fn data_exists(path: &Path) -> bool {
+        if !path.exists() || !path.is_dir() {
+            return false;
+        }
+        fs::read_dir(&path).unwrap().next().is_some()
+    }
+
+    fn remove_dir_safe(path: &Path) {
+        if path.exists() {
+            let trash = path.with_extension("REMOVE");
```

Comment:
Actually, this can leave dirty data forever if it's renamed but failed to remove.